Implements coordinator pattern for Growatt component data fetching

johanzander opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
This PR implements the data coordinator pattern for the Growatt component integration. This refactoring centralizes data fetching, error handling, and state management without changing existing functionality. Once this PR is accepted I will add growatt number and switch entities for which coordinator pattern is a pre-requisite.


## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests
- [X] Refactoring of existing code that enables new features in secondary step
 
## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [X] The code change is tested and works locally.
- [X] Local tests pass. **Your PR cannot be merged unless tests pass**
- [X] There is no commented out code in this PR.
- [X] I have followed the [development checklist][dev-checklist]
- [X] I have followed the [perfect PR recommendations][perfect-pr]
- [X] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

johanzander committed:
Refactors Growatt integration: implements coordinator pattern for data fetching and sensor management

johanzander committed:
Merge branch 'dev' into growatt_coordinator

johanzander committed:
Merge branch 'dev' into growatt_coordinator

johanzander commented:
Any chance to have a look at this PR?

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -1,21 +1,92 @@
 
-from .const import PLATFORMS
+    # Store the coordinator in hass.data
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][config_entry.entry_id] = {
+        "total": total_coordinator,
+        "devices": {},
+    }
```
Now that we start storing this, we should rather extend the `ConfigEntry` type and store this in `entry.runtime_data`. For the type we can use a dataclass.

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+_LOGGER = logging.getLogger(__name__)
+
+
+class GrowattCoordinator(DataUpdateCoordinator):
```
```suggestion
class GrowattCoordinator(DataUpdateCoordinator[..]):
```
Can we type the return type of `_async_update_data`?

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+                )
+                del total_info["deviceList"]
+                plant_money_text, currency = total_info["plantMoneyText"].split("/")
+                total_info["plantMoneyText"] = plant_money_text
+                total_info["currency"] = currency
+                self.data = total_info
+            elif self.device_type == "inverter":
+                inverter_info = await self.hass.async_add_executor_job(
+                    self.api.inverter_detail, self.device_id
+                )
+                self.data = inverter_info
+            elif self.device_type == "tlx":
+                tlx_info = await self.hass.async_add_executor_job(
+                    self.api.tlx_detail, self.device_id
+                )
+                self.data = tlx_info["data"]
+            elif self.device_type == "storage":
+                storage_info_detail = await self.hass.async_add_executor_job(
+                    self.api.storage_params, self.device_id
+                )
+                storage_energy_overview = await self.hass.async_add_executor_job(
+                    self.api.storage_energy_overview, self.plant_id, self.device_id
+                )
+                self.data = {
+                    **storage_info_detail["storageDetailBean"],
+                    **storage_energy_overview,
+                }
+            elif self.device_type == "mix":
+                mix_info = await self.hass.async_add_executor_job(
+                    self.api.mix_info, self.device_id
+                )
+                mix_totals = await self.hass.async_add_executor_job(
+                    self.api.mix_totals, self.device_id, self.plant_id
+                )
+                mix_system_status = await self.hass.async_add_executor_job(
+                    self.api.mix_system_status, self.device_id, self.plant_id
+                )
+                mix_detail = await self.hass.async_add_executor_job(
+                    self.api.mix_detail, self.device_id, self.plant_id
+                )
```
So, switching from the event loop to an executor job is expensive, so we try to avoid running multiple after each other. So I think the easiest for now (to make the code change clear), is to just move the content of the current `_async_update_data` and move it to a sync function and then make `_async_update_data` run that new function in the executor job. This way we are more efficient

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+            update_interval=SCAN_INTERVAL,
+        )
+
+    async def _async_update_data(self):
```
return type

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+        self.data: dict[str, Any] = {}
+        self.previous_values: dict[str, Any] = {}
+
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=f"{DOMAIN} ({device_id})",
+            update_interval=SCAN_INTERVAL,
+        )
```
we should input the config entry in here

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+        self.device_id = device_id
+        self.device_type = device_type
+        self.plant_id = plant_id
+        self.data: dict[str, Any] = {}
```
Let's not preset self.data yet

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -1,21 +1,92 @@
+    devices, plant_id = await hass.async_add_executor_job(get_device_list, api, config)
+
+    # Create a coordinator for the total sensors
+    total_coordinator = GrowattCoordinator(
+        hass, config_entry, plant_id, "total", plant_id
+    )
 
-from .const import PLATFORMS
+    # Store the coordinator in hass.data
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][config_entry.entry_id] = {
+        "total": total_coordinator,
+        "devices": {},
+    }
 
+    # Create coordinators for each device
+    for device in devices:
+        device_type = device["deviceType"]
+        if device_type in ["inverter", "tlx", "storage", "mix"]:
+            device_coordinator = GrowattCoordinator(
+                hass, config_entry, device["deviceSn"], device_type, plant_id
+            )
+            hass.data[DOMAIN][config_entry.entry_id]["devices"][device["deviceSn"]] = (
+                device_coordinator
+            )
```
We should execute the `async_config_entry_first_refresh` of the coordinators. This will test that the coordinator is able to retrieve data and then the entities are able to get a state upfront. If I have to guess, currently your entities do not have a state until the first update

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+                self.device_id,
+                self.device_type,
+            )
+        except json.decoder.JSONDecodeError:
+            _LOGGER.error("Unable to fetch data from Growatt server")
```
If we can't fetch data for whatever reason, we should raise `UpdateFailed`. If this happens during startup (using `async_config_entry_first_refresh`) the config entry will retry later. If this happens during runtime, the attached entities will go unavailable (which is correct as we don't get any info for it)

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,215 @@
+            )
+            diff = float(api_value) - float(previous_value)
+
+            # Check if the value has dropped (negative value i.e. < 0) and it has only
+            # dropped by a small amount, if so, use the previous value.
+            # Note - The energy dashboard takes care of drops within 10%
+            # of the current value, however if the value is low e.g. 0.2
+            # and drops by 0.1 it classes as a reset.
+            if -(entity_description.previous_value_drop_threshold) <= diff < 0:
+                _LOGGER.debug(
+                    (
+                        "Diff is negative, but only by a small amount therefore not a"
+                        " nightly reset, using previous value (%s) instead of api value"
+                        " (%s)"
+                    ),
+                    previous_value,
+                    api_value,
+                )
+                return_value = previous_value
+            else:
+                _LOGGER.debug(
+                    "%s - No drop detected, using API value", entity_description.name
+                )
+
+        # Lifetime total values should always be increasing, they will never reset,
+        # however the API sometimes returns 0 values when the clock turns to 00:00
+        # local time in that scenario we should just return the previous value
+        if entity_description.never_resets and api_value == 0 and previous_value:
+            _LOGGER.debug(
+                (
+                    "API value is 0, but this value should never reset, returning"
+                    " previous value (%s) instead"
+                ),
+                previous_value,
+            )
+            return_value = previous_value
+
+        self.previous_values[variable] = return_value
+
+        return return_value
```
Okay so since this already was here before I am fine with this, however, I would recommend to keep the logic on how entities get their data in the platform and not in the coordinator, as the coordinator is for fetching data and doesn't care about how entities show it.

But since this was already there, I would recommend to look at this in a followup

joostlek commented on homeassistant/components/growatt_server/sensor/__init__.py:
```diff
@@ -34,99 +22,56 @@
+
+    # Add total sensors
+    total_coordinator = coordinators["total"]
+    entities.extend(
+        [
+            GrowattSensor(
+                total_coordinator,
+                name=f"{config_entry.data['name']} Total",
+                serial_id=config_entry.data["plant_id"],
+                unique_id=f"{config_entry.data['plant_id']}-{description.key}",
+                description=description,
+            )
+            for description in TOTAL_SENSOR_TYPES
+        ]
+    )
```
```suggestion
    entities.extend(
        GrowattSensor(
            total_coordinator,
            name=f"{config_entry.data['name']} Total",
            serial_id=config_entry.data["plant_id"],
            unique_id=f"{config_entry.data['plant_id']}-{description.key}",
            description=description,
        )
        for description in TOTAL_SENSOR_TYPES
    )
```

joostlek commented on homeassistant/components/growatt_server/sensor/__init__.py:
```diff
@@ -136,33 +81,38 @@ async def async_setup_entry(
     async_add_entities(entities, True)
```
```suggestion
    async_add_entities(entities)
```
After config_entry_first_refresh we already have a state

joostlek commented on homeassistant/components/growatt_server/sensor/__init__.py:
```diff
@@ -136,33 +81,38 @@ async def async_setup_entry(
 
     _attr_has_entity_name = True
-
+    coordinator: GrowattCoordinator
```
```suggestion
```

joostlek replied:
Still relevant

johanzander committed:
Merge branch 'home-assistant:dev' into growatt_coordinator

johanvolvo reviewed (commented):
Summary
This PR addresses the review comments from @joostlek on #143373 for the Growatt Server integration. The following changes have been made to align with Home Assistant best practices and reviewer feedback:

Changes

Runtime Data Storage:
Moved coordinator and device data storage from hass.data to config_entry.runtime_data, using a dataclass for type safety and clarity.

Coordinator Initialization:
Coordinators now execute async_config_entry_first_refresh during setup to ensure entities have a state immediately after setup.

Type Hints and Typing Improvements:
Added explicit type hints, including the return type for _async_update_data in the coordinator.
Updated GrowattSensor to inherit from CoordinatorEntity[GrowattCoordinator] for better typing.

Data Fetching Efficiency:
Refactored coordinator data fetching: moved the main data-fetching logic to a synchronous function, and now call it from _async_update_data using a single executor job, reducing event loop switches.

Error Handling:
If data fetching fails, the coordinator now raises UpdateFailed to ensure proper retry and entity availability behavior.

Other Improvements:
Avoided presetting self.data in the coordinator constructor.
Updated async_add_entities usage in the sensor platform to not force an immediate update, as the state is already available after the first refresh.

Reviewer Notes
The logic for handling previous values and drop thresholds remains in the coordinator for now, as discussed, but can be moved to the platform/entities in a follow-up PR if desired.
All changes follow the Home Assistant code style and typing guidelines.

johanzander committed:
growatt_server: address review feedback, improve coordinator pattern

- Store runtime data in config_entry.runtime_data using dataclass
- Run async_config_entry_first_refresh for coordinators
- Add type hints and improve typing
- Refactor data fetching to minimize executor jobs
- Raise UpdateFailed on fetch errors
- Minor code style and efficiency improvements

johanzander commented:
**Summary**

I've now adressed the review comments from @joostlek on #143373 for the Growatt Server integration. The following changes have been made to align with Home Assistant best practices and reviewer feedback:

**Changes**
* Runtime Data Storage:
    * Moved coordinator and device data storage from hass.data to config_entry.runtime_data, using a dataclass for type safety and clarity.

* Coordinator Initialization:
    * Coordinators now execute async_config_entry_first_refresh during setup to ensure entities have a state immediately after setup.

* Type Hints and Typing Improvements:
    * Added explicit type hints, including the return type for _async_update_data in the coordinator.
    * Updated GrowattSensor to inherit from CoordinatorEntity[GrowattCoordinator] for better typing.

* Data Fetching Efficiency:
    * Refactored coordinator data fetching: moved the main data-fetching logic to a synchronous function, and now call it from _async_update_data using a single executor job, reducing event loop switches.

* Error Handling:
    * If data fetching fails, the coordinator now raises UpdateFailed to ensure proper retry and entity availability behavior.

* Other Improvements:
    * Avoided presetting self.data in the coordinator constructor.
    * Updated async_add_entities usage in the sensor platform to not force an immediate update, as the state is already available after the first refresh.

**Reviewer Notes**
* The logic for handling previous values and drop thresholds remains in the coordinator for now, as discussed, but can be moved to the platform/entities in a follow-up PR if desired.
* All changes follow the Home Assistant code style and typing guidelines.

johanzander committed:
Merge branch 'dev' into growatt_coordinator

johanzander committed:
Merge branch 'dev' into growatt_coordinator

johanzander committed:
Merge branch 'dev' into growatt_coordinator

joostlek reviewed (changes requested):
There still are some comments open and unresponded to

johanzander committed:
Merge branch 'home-assistant:dev' into growatt_coordinator

johanzander committed:
Adresses review comments

johanzander commented:
@joostlek - All requested changes should be adressed by now, let me know if you have any further comments!

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -1,21 +1,98 @@
+    return devices, plant_id
+
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
```
Let's extend the `ConfigEntry` to type the `entry.runtime_data` field. Example can be found here https://developers.home-assistant.io/docs/core/integration-quality-scale/rules/runtime-data

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -1,21 +1,98 @@
+
+async def async_setup_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
+    """Set up Growatt from a config entry."""
+    config = {**config_entry.data}
```
```suggestion
    config = config_entry.data
```

joostlek commented on homeassistant/components/growatt_server/__init__.py:
```diff
@@ -1,21 +1,98 @@
 
 
-async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+async def async_unload_entry(hass: HomeAssistant, config_entry: ConfigEntry) -> bool:
```
use the extended config entry here too

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,206 @@
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        config_entry: ConfigEntry,
```
here too

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,206 @@
+        plant_id: str,
+    ) -> None:
+        """Initialize the coordinator."""
+        self.config = {**config_entry.data}
```
I am not sure why we need to store this.

Let's overwrite the config entry type on the coordinator, then we can just freely access `coordinator.config_entry.data`

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,206 @@
+
+
+class GrowattCoordinator(DataUpdateCoordinator[dict[str, Any]]):
+    """Coordinator to manage Growatt data fetching."""
+
+    def __init__(
```
```suggestion
    """Coordinator to manage Growatt data fetching."""

    config_entry: ConfigEntry

    def __init__(
```
But also please use the extended config entry here

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,206 @@
+            add_random_user_id=True, agent_identifier=self.username
+        )
+
+        # Store the ConfigEntry object
+        self.config_entry = config_entry
```
this is already happening in the super constructor

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,206 @@
+            _LOGGER.error("Unable to fetch data from Growatt server: %s", err)
+            raise UpdateFailed(f"Error fetching data: {err}") from err
+
+    def get_currency(self):
+        """Get the currency."""
+        return self.data.get("currency")
```
Why is this a function?

johanzander replied:
good question, but i'm not the original author, so I don't know - I'm just trying to refactor to use coordinator and the function was there already.

joostlek commented on homeassistant/components/growatt_server/coordinator.py:
```diff
@@ -0,0 +1,206 @@
+        """Get the currency."""
+        return self.data.get("currency")
+
+    def get_data(self, entity_description):
```
can we add typing and return typing?

joostlek commented on homeassistant/components/growatt_server/sensor/__init__.py:
```diff
@@ -34,135 +22,94 @@
 
         self._attr_device_info = DeviceInfo(
-            identifiers={(DOMAIN, probe.device_id)},
+            identifiers={(DOMAIN, serial_id)},
```
Is this the same identifier?

johanzander committed:
Refactor Growatt integration to use GrowattConfigEntry instead of ConfigEntry

johanzander committed:
Refactor get_device_list and async_setup_entry to use Mapping for config type and improve URL handling

johanzander committed:
Adds type hints to get_data()

johanzander committed:
Merge branch 'dev' into growatt_coordinator

joostlek reviewed (approved):
It would be awesome if you could maybe look into stepping up as codeowner as the integration is still quite used and deserves more love.
